### PR TITLE
Add setting for adjustment of audio volume above 100 per cent:  Part 2

### PR DESCRIPTION
**Pull request by thesquash**
There is often a need for the user to increase the audio playback volume above
the volume level known as "100% volume".  While increasing the audio volume
above 100% can result in degraded audio quality, sometimes the audio was, for
example, originally recorded at an extremely low volume, and the user has no
other option to clearly hear the audio.  Unfortunately, most MATE applications
with volume controls do not allow the user to set the volume level above 100%.
For example, the main MATE Sound Preferences dialog lets you set the audio
volume beyond 100% (when possible), whereas the Volume Control Applet, Volume
Control status icon, and special "multimedia" volume control keys do not.  In
fact, if the user even tries to change the volume using any of the latter
methods, and the current volume level is above 100%, these latter methods will
all reduce the volume to 100%, even if the user tried to increase the volume!

This is part 2 of a patch to change this situation.  This patch adds this
capability to the MATE Volume Control Applet and the MATE Volume Control
status icon.  It also adds a check box to the MATE Volume Control Preferences
dialog to enable or disable this capability, and for consistency, the main
volume control at the top of that dialog also obeys the current setting of
that check box.  This commit is by far the largest one, and is the main part
of the patch.

Part 1:  https://github.com/mate-desktop/mate-desktop/pull/530
Part 3:  https://github.com/mate-desktop/mate-settings-daemon/pull/389

When all three parts are applied, https://github.com/mate-desktop/mate-desktop/issues/246 can be closed.

### CODE REVIEW

File: mate-volume-control/gvc-channel-bar.c [hunk 1]
```diff
@@ -369,18 +401,11 @@ update_adjustment_value (GvcChannelBar *bar)
...
-        gdouble range = maximum - minimum;
-
-        /* round value to nearest hundreth of the range */
-        gdouble new_value = minimum + round (((value - minimum) / range) * 100) * (range / 100);
-
         g_signal_handlers_block_by_func (G_OBJECT (bar->priv->adjustment),
                                          on_adjustment_value_changed,
                                          bar);
 
-        gtk_adjustment_set_value (bar->priv->adjustment, new_value);
```
Comment (rbuj) [inline mate-volume-control/gvc-channel-bar.c:383]: It should go in a separate commit as it's unrelated to the commit title. See #192 which can be closed once you've pushed the change into another PR. Keep in mind that it's a bug fix and you shouldn't mix it with a new functionality.

File: mate-volume-control/gvc-channel-bar.c [hunk 2]
```diff
@@ -286,6 +288,28 @@ update_layout (GvcChannelBar *bar)
...
+{
+        gdouble normal_volume;
+        gdouble maximum_volume;
+        int     calculated_scale_size = SCALE_SIZE;
+
+        if (bar->priv->extended && bar->priv->control != NULL)
+        {
+                normal_volume = mate_mixer_stream_control_get_normal_volume (bar->priv->control);
+                maximum_volume = mate_mixer_stream_control_get_max_volume (bar->priv->control);
+                calculated_scale_size = (maximum_volume / normal_volume) * SCALE_SIZE;
```
Comment (rbuj) [inline mate-volume-control/gvc-channel-bar.c:302]: here you can have a problem of loss of precision and sing. int = (guint / guint) * int

File: mate-volume-control/gvc-mixer-dialog.c [hunk 3]
```diff
@@ -1972,11 +1976,26 @@ gvc_mixer_dialog_constructor (GType                  type,
         gvc_channel_bar_set_name (GVC_CHANNEL_BAR (self->priv->output_bar),
                                   _("_Output volume: "));
 
+        g_settings_bind (self->priv->sound_settings, "volume-overamplifiable",
+                         self->priv->output_bar,     "extended",
+                         G_SETTINGS_BIND_GET);
+
```
Comment (rbuj) [inline mate-volume-control/gvc-mixer-dialog.c:1982]: update create_bar
```C
static GtkWidget *
create_bar (GvcMixerDialog *dialog, gboolean use_size_group, gboolean symmetric)
{
        GtkWidget *bar;

        bar = gvc_channel_bar_new (NULL);

        if (use_size_group == TRUE)
                gvc_channel_bar_set_size_group (GVC_CHANNEL_BAR (bar),
                                                dialog->priv->size_group,
                                                symmetric);

        g_object_set (G_OBJECT (bar),
                      "orientation", GTK_ORIENTATION_HORIZONTAL,
                      "show-mute",   TRUE,
                      "show-icons",  TRUE,
                      "show-marks",  TRUE, NULL);
        return bar;
}
```

File: mate-volume-control/gvc-stream-applet-icon.c [hunk 4]
```diff
@@ -731,6 +734,16 @@ gvc_stream_applet_icon_init (GvcStreamAppletIcon *icon)
         gvc_channel_bar_set_orientation (GVC_CHANNEL_BAR (icon->priv->bar),
                                          GTK_ORIENTATION_VERTICAL);
 
+        gvc_channel_bar_set_show_marks (GVC_CHANNEL_BAR (icon->priv->bar),
+                                        TRUE);
```
Comment (rbuj) [inline mate-volume-control/gvc-stream-applet-icon.c:738]:
```C
g_settings_bind ("icon->priv->sound_settings, "volume-overamplifiable",
                 icon->priv->bar, "show-marks",
                 G_SETTINGS_BIND_GET);

File: mate-volume-control/gvc-stream-status-icon.c [hunk 5]
```diff
@@ -745,6 +748,16 @@ gvc_stream_status_icon_init (GvcStreamStatusIcon *icon)
         gvc_channel_bar_set_orientation (GVC_CHANNEL_BAR (icon->priv->bar),
                                          GTK_ORIENTATION_VERTICAL);
 
+        gvc_channel_bar_set_show_marks (GVC_CHANNEL_BAR (icon->priv->bar),
+                                        TRUE);
```
Comment (rbuj) [inline mate-volume-control/gvc-stream-status-icon.c:752]:
```C
g_settings_bind (icon->priv->sound_settings, "volume-overamplifiable",
                 icon->priv->bar, "show-marks",
                 G_SETTINGS_BIND_GET);
```

File: mate-volume-control/gvc-mixer-dialog.c [hunk 6]
```diff
@@ -1972,11 +1976,26 @@ gvc_mixer_dialog_constructor (GType                  type,
...
+        self->priv->volume_overamplifiable_button =
+                gtk_check_button_new_with_mnemonic (_("Allow output volume to e_xceed 100%"));
+
+        g_settings_bind (self->priv->sound_settings,                "volume-overamplifiable",
+                         self->priv->volume_overamplifiable_button, "active",
+                         G_SETTINGS_BIND_GET|G_SETTINGS_BIND_SET);
+
+        gtk_box_pack_start (GTK_BOX (self->priv->output_stream_box),
+                            self->priv->volume_overamplifiable_button,
+                            TRUE, TRUE, 0);
```
Comment (rbuj) [inline mate-volume-control/gvc-mixer-dialog.c:1998]: the box should not fill expand % fill horizontally:

Captura de pantalla 2022-08-03 a les 23 58 23

```C
        gtk_box_pack_start (GTK_BOX (self->priv->output_stream_box),
                            self->priv->volume_overamplifiable_button,
                            FALSE, FALSE, 0);
```

File: mate-volume-control/gvc-mixer-dialog.c [hunk 7]
```diff
@@ -1972,11 +1976,26 @@ gvc_mixer_dialog_constructor (GType                  type,
...
+
         gtk_widget_show (self->priv->output_bar);
         gtk_widget_set_sensitive (self->priv->output_bar, FALSE);
 
         gtk_box_pack_start (GTK_BOX (self->priv->output_stream_box),
-                            self->priv->output_bar, TRUE, TRUE, 12);
+                            self->priv->output_bar, TRUE, TRUE, 0);
+
+        self->priv->volume_overamplifiable_button =
+                gtk_check_button_new_with_mnemonic (_("Allow output volume to e_xceed 100%"));
```
Comment (rbuj) [inline mate-volume-control/gvc-mixer-dialog.c:1990]: This check button is only valid when the backend is PA and the it has a sw mixer (`flags & MATE_MIXER_STREAM_CONTROL_HAS_DECIBEL`), maybe it should be disabled for the other backends (OSS, alsa) and when PA has a hw mixer.

File: mate-volume-control/gvc-channel-bar.c (hunk 1 above)
Comment (thesquash) [reply mate-volume-control/gvc-channel-bar.c]: The "unrelated" code has been removed from this PR.

File: mate-volume-control/gvc-channel-bar.c (hunk 2 above)
Comment (thesquash) [reply mate-volume-control/gvc-channel-bar.c:302]: Except `normal_volume` and `maximum_volume` are `gdouble`s, not `guint`s.  That basic code is copied from another place above and nearby, where `gdouble`s are used too.  And, this code *does* verifiably work.

File: mate-volume-control/gvc-mixer-dialog.c (hunk 6 above)
Comment (thesquash) [reply mate-volume-control/gvc-mixer-dialog.c]: Done.

File: mate-volume-control/gvc-mixer-dialog.c (hunk 7 above)
Comment (thesquash) [reply mate-volume-control/gvc-mixer-dialog.c]: This has now been dealt with.  Image below shows the MATE Sound Preferences with PulseAudio SW mixer (left) and plain old ALSA (right).

MATE-Mixer-with-and-without-overamplification-support-in-audio-backend

**Comment by thesquash** [comment]
I think I've resolved or at least addressed all issues above.

**Comment by rbuj** [comment]
sorry, due to disagreements with the manager of the MATE team I'm no longer part of it, I hope another team member will review your changes, regards.

https://github.com/mate-desktop/atril/pull/570#issuecomment-1208630425

**Comment by raveit65** [comment]
This isn't true, because MATE doesn't have a manager, everybody from the team has equal rights.
And no one did say to you to leave the team.
It was your own decision to drop your privileges..... 
....... in the end only because you didn't want to write a little mail with a sentence of 5 words to travis-CI to trigger an automatic process.
Sounds childish for me.

**Comment by thesquash** [comment]
Man, that's such a darn shame, that he left.  I really liked working with him.  He had a good eye for my dumb mistakes, and I respect him for that and his other development work.  Still, I respect his decision to leave.

---

We've already de-railed this pull request a little bit by talking about his departure instead of the code, so this offers me a little opportunity.  I'd hate for this to sound like I'm just taking advantage of this predicament, but I don't know where else or when else to ask this:

I've wanted to formally join the MATE Project for quite some time now.  I've contributed code (off-and-on) for over two years now.  My three questions are:

 - What would my responsibilities be as an official team member?
 - Does anybody think I'm unfit for entry to the team (I'm too stupid, etc)?
 - What would it take for me to join, if I agree to fulfill my responsibilities as a team member?

Thank you.  I'd greatly appreciate a response.

**Comment by lukefromdc** [comment]
I'v seen your contributions over the years and would have no problem with adding you to the tean

On 8/10/2022 at 10:28 PM, "Gordon N. Squash" ***@***.***> wrote:
>
>Man, that's such a darn shame, that he left.  I really liked 
>working with him.  He had a good eye for my dumb mistakes, and I 
>respect him for that and his other development work.  Still, I 
>respect his decision to leave.
>
>---
>
>We've already de-railed this pull request a little bit by talking 
>about his departure instead of the code, so this offers me a 
>little opportunity.  I'd hate for this to sound like I'm just 
>taking advantage of this predicament, but I don't know where else 
>or when else to ask this:
>
>I've wanted to formally join the MATE Project for quite some time 
>now.  I've contributed code (off-and-on) for over two years now.  
>My three questions are:
>
> - What would my responsibilities be as an official team member?
> - Does anybody think I'm unfit for entry to the team (I'm too 
>stupid, etc)?
> - What would it take for me to join, if I agree to fulfill my 
>responsibilities as a team member?
>
>Thank you.  I'd greatly appreciate a response.
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/mate-desktop/mate-media/pull/191#issuecomment-
>1211491086
>You are receiving this because you are subscribed to this thread.
>
>Message ID: <mate-desktop/mate-
***@***.***>

**Comment by thesquash** [comment]
Well, thank you, but is there somewhere I can read about what would be expected from me if I joined?  There's obviously routine things -- like the CI maintenance that ultimately led to Buj leaving -- and there's general quality assurance and a sharp eye for defects.  But there's got to be something else to the job, right?

Whatever the job entails, I'm almost certain I can do it.  I just want to know what I'm getting into, before I'm up to my neck in it.  TIA.

**Comment by lukefromdc** [comment]
As for the PR itself, is it ready to go or more work needed? I can merge it when it is ready

**Comment by raveit65** [comment]
> There's obviously routine things -- like the CI maintenance

This is the hardest hard job ever....
1. You need an account at travis-ci.com
2. Add Mate Desktop organization to your account.
3. You will see a page like that
grafik
4. When credits are at 1000 write a mail like this
```
Dear Travis support,

i want to noticed you that sponsored credits are running out. Can you please update credits in result we can continue to use CI ?

Thank you very much for your sponsoring.

Best regards

Wolfgang 
```
5. After 2-3 hours you will receive a mail.
Bildschirmfoto zu 2022-08-12 09-04-39
That's it.

**Comment by raveit65** [comment]
> I've wanted to formally join the MATE Project for quite some time now. I've contributed code (off-and-on) for over two years now. 

+1

**Comment by raveit65** [comment]
>  There's obviously routine things

Well, i am leaving Mate since several months due a lack of time and maintaining Mate is since 2012 is a long time, and i don't get any motivation any more.
The reason that i was a little bit active the last weeks is to support Mate desktop for Rocky Linux from Fedora side.
Rocky Linux is the replacement for CentOs which we support with fedora epel repo.
So i did a few new 1.26 releases and a rebase of caja-1.26 with lower glib2 requirement.

So, this work is done now and i redefine my role as fedora distro maintainer only!

A lot of tasks which i had done in the background are looking for a new maintainer here.
- Transifex translations for Mate desktop environment and website needs an admin.
- pulling translations from transifex for a new release and checking that new translation doesn't break building. Sadly translators do a lot of syntax mistakes.........
- Releasing new packages from time to time.
- Administration of Mate desktop website
- Administration of github repo and travis CI
- Maintaining mate-themes

So, this needs to be done by the team until now.
Of course I can give a little briefing if anyone is interested.
Good luck

PS: I don't mind if some like to work again with other devs who had leave the project.

**Comment by thesquash** [comment]
> As for the PR itself, is it ready to go or more work needed?

Well, I don't see that it needs any more work, myself.  I think it can be merged, but I'm not part of the team yet so my word here isn't worth too much.

---

@raveit65:  The tasks and responsibilities that you listed do not appear to be terribly difficult, and they seem that they won't absorb masses of my time either.  So if everybody is on board with it, I want to join the team.

---

Thank you all in advance.

**Comment by thesquash** [comment]
Thanks for inviting me to the team, @raveit65.  (Obviously I accepted the invitation.)

---

Back to the topic of this pull request, though:  Does anybody disagree with this PR, or may I merge it?

**Review by raveit65 — APPROVED** [review]
LGTM

**Comment by raveit65** [comment]
What i really like is that this change works also for volume slider of the applet. I didn't saw that in my last test.
Thanks

**Comment by zhuyaliang** [comment]
@thesquash Welcome to this team
